From: Mike Snitzer <snitzer@redhat.com>
Date: Tue, 16 Feb 2010 16:27:55 -0500
Subject: [block] fix bio_add_page for non trivial merge_bvec_fn case
Message-id: <1266337675-23798-1-git-send-email-snitzer@redhat.com>
Patchwork-id: 23285
O-Subject: [RHEL6 PATCH] block: fix bio_add_page for non trivial merge_bvec_fn
	case
Bugzilla: 565890
RH-Acked-by: Josef Bacik <josef@redhat.com>
RH-Acked-by: Jerome Marchand <jmarchan@redhat.com>

Upstream in both linux-2.6 and linux-2.6.32.y

Bugzilla #565890

commit 1d6165851cd8e3f919d446cd6da35dee44e8837e
Author: Dmitry Monakhov <dmonakhov@openvz.org>
Date:   Wed Jan 27 22:44:36 2010 +0300

    block: fix bio_add_page for non trivial merge_bvec_fn case

    We have to properly decrease bi_size in order to merge_bvec_fn return
    right result.  Otherwise this result in false merge rejects for two
    absolutely valid bio_vecs.  This may cause significant performance
    penalty for example fs_block_size == 1k and block device is raid0 with
    small chunk_size = 8k. Then it is impossible to merge 7-th fs-block in
    to bio which already has 6 fs-blocks.

    Cc: <stable@kernel.org>
    Signed-off-by: Dmitry Monakhov <dmonakhov@openvz.org>
    Signed-off-by: Jens Axboe <jens.axboe@oracle.com>

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/fs/bio.c b/fs/bio.c
index 12da5db..e0c9e71 100644
--- a/fs/bio.c
+++ b/fs/bio.c
@@ -542,13 +542,18 @@ static int __bio_add_page(struct request_queue *q, struct bio *bio, struct page
 
 		if (page == prev->bv_page &&
 		    offset == prev->bv_offset + prev->bv_len) {
+			unsigned int prev_bv_len = prev->bv_len;
 			prev->bv_len += len;
 
 			if (q->merge_bvec_fn) {
 				struct bvec_merge_data bvm = {
+					/* prev_bvec is already charged in
+					   bi_size, discharge it in order to
+					   simulate merging updated prev_bvec
+					   as new bvec. */
 					.bi_bdev = bio->bi_bdev,
 					.bi_sector = bio->bi_sector,
-					.bi_size = bio->bi_size,
+					.bi_size = bio->bi_size - prev_bv_len,
 					.bi_rw = bio->bi_rw,
 				};
 
